Add the private prefetch proxy blog post (#2797)
GoogleChrome/developer.chrome.com: site/en/blog/private-prefetch-proxy/index.md

---
# Required
layout: 'layouts/blog-post.njk'
title: Private prefetch proxy in Chrome
description: >

>   Speeding up Largest Contentful Paint with cross-site prefetching.

Reviewer: ```suggestion
  Speeding up Largest Contentful Paint (LCP) with cross-site prefetching.
```

subhead: >

>   Speeding up Largest Contentful Paint with cross-site prefetching

Reviewer: ```suggestion
  Speeding up Largest Contentful Paint (LCP) with cross-site prefetching.
```

date: 2022-05-11
authors:
  - katiehempenius
  - kenjibaheux
  - buettner
tags:
  - privacy
  - security
hero: 'image/kheDArv5csY6rvQUJDbWRscckLr1/MKR7llxWeUvVif0CDGPM.jpg'
alt: >
  A gate with a sign reading private.
---

> Since Chrome 92, we've been running an early access program (EAP) for a feature called private prefetch proxy, which speeds up out-going navigations from Google Search by 30% at the median. This private prefetch proxy feature allows the prefetching of cross-origin content without exposing user information to the destination website until the user navigates. We expect that the feature will graduate from this EAP as early as Chrome 103, thereby no longer requiring websites to opt-in, and allowing other referrer websites to safely speed up cross-site navigations.

Reviewer: It wasn't clear to me what the proper case was for Private Prefetch Proxy, but it's inconsistent in other parts of the doc from what I've seen.

Read on to learn about [how this feature works](#how), 

> [how it can help significantly improve your sites' Largest Contentful Paint](#owners), 

Reviewer: ```suggestion
[how it can help significantly improve your sites' Largest Contentful Paint (LCP)](#owners), 
```

or [how referrer websites can help their users](#referrers) achieve their goals by speeding up cross-site navigations.

> ## How private prefetch proxy works {: #how }

Reviewer: ```suggestion
## How Private Prefetch Proxy works {: #how }
```

### Secure communication channel

> This feature uses a [CONNECT](https://tools.ietf.org/html/rfc7231#section-4.3.6) proxy to establish a secure communication channel between Chrome and the server hosting the content to be prefetched. This secure communication channel prevents the proxy from inspecting any data transfer. Notably, while the private prefetch proxy necessarily sees the host name in order to establish a secure communication channel, it does not see the full URLs, nor the resources themselves.

Reviewer: ```suggestion
This feature uses a [`CONNECT`](https://tools.ietf.org/html/rfc7231#section-4.3.6) proxy to establish a secure communication channel between Chrome and the server hosting the content to be prefetched. This secure communication channel prevents the proxy from inspecting any data transfer. Notably, while the private prefetch proxy necessarily sees the host name in order to establish a secure communication channel, it does not see the full URLs, nor the resources themselves.
```

Reviewer 2: That's fine by me.

> This feature uses a [CONNECT](https://tools.ietf.org/html/rfc7231#section-4.3.6) proxy to establish a secure communication channel between Chrome and the server hosting the content to be prefetched. This secure communication channel prevents the proxy from inspecting any data transfer. Notably, while the private prefetch proxy necessarily sees the host name in order to establish a secure communication channel, it does not see the full URLs, nor the resources themselves.

Reviewer: ```suggestion
This feature uses a [CONNECT](https://tools.ietf.org/html/rfc7231#section-4.3.6) proxy to establish a secure communication channel between Chrome and the server hosting the content to be prefetched. This secure communication channel prevents the proxy from inspecting any data transfer. Notably, while Private Prefetch Proxy necessarily sees the host name in order to establish a secure communication channel, it does not see the full URLs, nor the resources themselves.
```

<figure>
{% Img src="image/kheDArv5csY6rvQUJDbWRscckLr1/UF6dpd6QvSzGXXlM9YWE.gif", alt="Animation showing flow of data through proxy.", width="600", height="365" %}

>   <figcaption>Prefetching websites via a CONNECT proxy prevents leaking user information.</figcaption>

Reviewer: ```suggestion
  <figcaption>Prefetching websites via a <code>CONNECT</code> proxy prevents leaking user information.</figcaption>
```

</figure>

In addition, because the secure communication channel is encrypted end-to-end, intermediaries can neither observe the host names, nor the content of the prefetched sites. Finally, the proxy inherently prevents the destination server from seeing the user's IP address.

### Preventing user identification

> Beyond the network aspects detailed earlier, we also need to prevent servers from identifying a user at prefetch time, via information previously stored on their device. To that end, Chrome currently restricts the usage of Private Prefetch Proxy to websites for which the user has no cookies or other local state. Here are the restrictions for prefetch requests made via the private prefetch proxy:

Reviewer: ```suggestion
Beyond the network aspects detailed earlier, we also need to prevent servers from identifying a user at prefetch time, via information previously stored on their device. To that end, Chrome currently restricts the usage of Private Prefetch Proxy to websites for which the user has no cookies or other local state. Here are the restrictions for prefetch requests made via Private Prefetch Proxy:
```

Reviewer: I think normalizing this helps. I get why "the private prefetch proxy" isn't capitalized, but this minor change improves consistency IMO.

- **Cookies:** Prefetch requests are not allowed to carry cookies.
  - If there is a cookie for a resource, Chrome will make an uncredentialed fetch but will not use the response (see later [Caching](#caching) section). 
  - Although responses to a prefetch request can include cookies, these cookies will only be saved if the user navigates to the prefetched page.

> - **Fingerprinting:** Other surfaces which could be used for fingerprinting are also adjusted. For example, the User-Agent header sent by the prefetch proxy only carries limited information.

Reviewer: ```suggestion
- **Fingerprinting:** Other surfaces which could be used for fingerprinting are also adjusted. For example, the `User-Agent` header sent by prefetch proxy only carries limited information.
```

> In the future, we hope to expand private prefetch proxy to links with cookies or local state while maintaining the same privacy characteristics. See the [What's next](#next) section for more details.

Reviewer: ```suggestion
In the future, we hope to expand Private Prefetch Proxy to links with cookies or local state while maintaining the same privacy characteristics. See the [What's next](#next) section for more details.
```

### Caching {: #caching }

> Chrome will prefetch resources even if they are already in the cache, but they will not carry any conditional headers such as Etags or If-Modified-Since (these contain server-set values which could be used for tracking even without cookies). This prefetching is done to prevent leaking a client's cache state to the prefetched website. In addition, Chrome will only commit a prefetched resource to the cache if the user decides to navigate to the prefetched website.

Reviewer: ```suggestion
Chrome will prefetch resources even if they are already in the cache, but they will not carry any conditional headers such as `ETag` or `If-Modified-Since` (these contain server-set values which could be used for tracking even without cookies). This prefetching is done to prevent leaking a client's cache state to the prefetched website. In addition, Chrome will only commit a prefetched resource to the cache if the user decides to navigate to the prefetched website.
```

## Getting started with private prefetch proxy

### For website owners {: #owners }

> For the duration of the early access Program, website owners interested in private prefetch proxy will need to deploy a traffic advice file on their server with the following content: 

Reviewer: ```suggestion
For the duration of the Early Access Program, website owners interested in Private Prefetch Proxy will need to deploy a traffic advice file on their server with the following content: 
```

```json

> [{
>     "user_agent": "prefetch-proxy",
>     "google_prefetch_proxy_eap": {
>     	"fraction": 1.0
>     }
> }]

Reviewer: ```suggestion
[{
  "user_agent": "prefetch-proxy",
  "google_prefetch_proxy_eap": {
    "fraction": 1.0
  }
}]
```

```

- The `google_prefetch_proxy_eap` section allows you to opt into the EAP.

> - The `fraction` field gives you control over the prefetch traffic by specifying how much the private prefetch proxy should let through. See the [traffic control](#traffic) section for more details.

Reviewer: ```suggestion
- The `fraction` field gives you control over the prefetch traffic by specifying how much Private Prefetch Proxy should let through. See the [traffic control](#traffic) section for more details.
```

> This traffic advice file should be placed under the `/.well-known/` path of your website, and served with a MIME type of `application/trafficadvice+json`. Doing so will allow Chrome to speed up navigations to your website from participating referrers on links for which the user has no cookies or local state. From past experiments, we've seen between 20% to 30% faster Largest Contentful Paint on prefetched navigations.

Reviewer: ```suggestion
This traffic advice file should be placed under the `/.well-known/` path of your website, and served with a MIME type of `application/trafficadvice+json`. Doing so will allow Chrome to speed up navigations to your website from participating referrers on links for which the user has no cookies or local state. From past experiments, we've seen between 20% to 30% faster [LCP](https://web.dev/lcp/) on prefetched navigations.
```

> In the future, we hope to expand this feature to links with cookies or local state while maintaining its privacy characteristics. The challenge with cookies is that they might be used to alter the user experience in hard to predict ways. So, website owners will most likely have to opt in or adjust their site to benefit from private prefetch proxy for links with cookies.

Reviewer: ```suggestion
In the future, we hope to expand this feature to links with cookies or local state while maintaining its privacy characteristics. The challenge with cookies is that they might be used to alter the user experience in hard to predict ways. So, website owners will most likely have to opt in or adjust their site to benefit from Private Prefetch Proxy for links with cookies.
```

Concretely, while the prefetch requests will remain uncredentialed, the web page will gain access to cookies, and other local state, when the user navigates to it. Developers could take advantage of this to add back the personalization and changes based on cookies or local state. Or perhaps, developers might also be interested in declaring certain resources as perfectly fine to prefetch and use as is, without cookies (that is, resources that don't depend on any cookies). Please have a look at the [What's next](#next) section to learn more and inform our plan.

### Geo-dependent content or services

> If your website behaves differently (for example, different content, or selective access) across markets based on the user's IP address, you may wonder how to handle private prefetch proxy's prefetch requests. It's important to know that private prefetch proxy is powered by several servers spread across the globe, and that the IP of the proxy will geolocate to the country from which the user initiated a prefetch.

Reviewer: ```suggestion
If your website behaves differently (for example, different content, or selective access) across markets based on the user's IP address, you may wonder how to handle Private Prefetch Proxy's prefetch requests. It's important to know that Private Prefetch Proxy is powered by several servers spread across the globe, and that the IP of the proxy will geolocate to the country from which the user initiated a prefetch.
```

So, with that in mind, here is what we recommend:

> 1. Identify prefetch requests from a private prefetch proxy by the presence of a Sec-Purpose: Prefetch; anonymous-client-ip HTTP header.

Reviewer: ```suggestion
1. Identify prefetch requests from Private Prefetch Proxy by the presence of a `Sec-Purpose: Prefetch; anonymous-client-ip` HTTP header.
```

> 1. Lookup the geolocation of the private prefetch proxy that issued the request via its IP address. See [this resource](https://www.gstatic.com/chrome/prefetchproxy/prefetch_proxy_geofeed) for an up-to-date list of rolled-out geographies and the corresponding IP addresses.

Reviewer: ```suggestion
1. Lookup the geolocation of the Private Prefetch Proxy that issued the request via its IP address. See [this resource](https://www.gstatic.com/chrome/prefetchproxy/prefetch_proxy_geofeed) for an up-to-date list of rolled-out geographies and the corresponding IP addresses.
```

1. Serve resources in accordance with the market attached to this particular geolocation.

### Traffic control {: #traffic }

> From past experiments, we know that this feature typically results in less than 2% extra requests for main resources (for example HTML documents). That said, if you are the cautious kind, you can use the traffic advice's fraction field to control how much the private prefetch proxy should let through. You can start with a small fraction such as 0.3 (that is,. 30%), and gradually increase it to 1.0 (that is, 100%).

Reviewer: ```suggestion
From past experiments, we know that this feature typically results in less than 2% extra requests for main resources (for example HTML documents). That said, if you are the cautious kind, you can use the traffic advice's fraction field to control how much Private Prefetch Proxy should let through. You can start with a small fraction such as 0.3 (that is, 30%), and gradually increase it to 1.0 (that is, 100%).
```

```json

> [{
>     "user_agent": "prefetch-proxy",
>     "google_prefetch_proxy_eap": {
>         	"fraction": 0.3
>     }
> }]

Reviewer: ```suggestion
[{
  "user_agent": "prefetch-proxy",
  "google_prefetch_proxy_eap": {
    "fraction": 0.3
  }
}]
```

```

> The `fraction` field gives you control over the prefetch traffic by specifying how much the private prefetch proxy should let through. The value is a float between 0.0 (no prefetch at all) and 1.0 (100% of the prefetch requests get through).

Reviewer: ```suggestion
The `fraction` field gives you control over the prefetch traffic by specifying how much Private Prefetch Proxy should let through. The value is a float between 0.0 (no prefetch at all) and 1.0 (100% of the prefetch requests get through).
```

> This _traffic advice file_ should be placed under the `/.well-known/` path of your website, and served with a MIME type of `application/trafficadvice+json`.

Reviewer: I think this advice was given earlier.

Reviewer 2: Yes, I also noted that on my side. We can remove the second mention.

> For more flexibility (for example, a sudden peak of heavy access), you may want to temporarily reject prefetch requests (`Sec-Purpose: Prefetch; anonymous-client-ip`) with a 503 status code, and by setting the `Cache-control: no-store` header on the response. You may also add the [Retry-After](https://tools.ietf.org/html/rfc7231#section-6.6.4) header to tell Chrome how long to wait before retrying prefetch requests.

Reviewer: ```suggestion
For more flexibility (for example, a sudden peak of heavy access), you may want to temporarily reject prefetch requests (`Sec-Purpose: Prefetch; anonymous-client-ip`) with a 503 status code, and by setting the `Cache-control: no-store` header on the response. You may also add the [`Retry-After`](https://tools.ietf.org/html/rfc7231#section-6.6.4) header to tell Chrome how long to wait before retrying prefetch requests.
```

## For referrer website owners {: #referrers }

> If you operate a website with lots of links to other websites, you may be interested in using the private prefetch proxy feature to speed up these cross-origin navigations. You will need to add [speculation rules](https://web.dev/speculative-prerendering/#in-browser-speculation-rules-for-prefetch-and-prerender) to your pages for Chrome to know which page you believe it should prefetch via the private prefetch proxy. Here is a simple example:

Reviewer: ```suggestion
If you operate a website with lots of links to other websites, you may be interested in using the Private Prefetch Proxy feature to speed up these cross-origin navigations. You will need to add [speculation rules](https://web.dev/speculative-prerendering/#in-browser-speculation-rules-for-prefetch-and-prerender) to your pages for Chrome to know which page you believe it should prefetch via Private Prefetch Proxy. Here is a simple example:
```

```html

> <script type="speculationrules">
> {
>   "prefetch": [
> "source": "list",
> "urls": ["https://example.com/index.html"],
> "requires": ["anonymous-client-ip-when-cross-origin"]}
>   ]
> }
> </script>

Reviewer: ```suggestion
<script type="speculationrules">
{
  "prefetch": [
    "source": "list",
    "urls": ["https://example.com/index.html"],
    "requires": ["anonymous-client-ip-when-cross-origin"]
  ]
}
</script>
```
